### ISSUE TITLE
Feature Suggestion: Ability to add custom entry handlers when loading egg files

### ISSUE THREAD

**Issue by lifelandman** [panda3d/panda3d#1581]
## Description
Egg files tend to have mixed usefulness compared to bam files depending on the programmer. I personally think that egg files may be a bit better for things like storing levels, due to the ability to hold references to external files. Furthermore, egg-files are text based, meaning that if a programmer wanted to support user-made custom levels, it would hypothetically be easier for the user to make small corrections to the levels if any of the provided tools didn't support such an adjustment.

One feature that I think would greatly improve egg files would be the ability to write handlers for custom entries, so that when the loader encounters an entry that it doesn't recognize, it can consult a list of set entry-handlers to process it. (similar to custom file handlers when building.) Essentially, the entry handler would receive a list of the entry's scalars and any nodes inside of it, and perform an operation on this data as the programmer specifies.

I'm aware that most use-cases for this could be solved using node tags, but I feel that as the complexity of any use-case increases, the legibility of the tags would decrease.

## Use Cases

1. Storing information in an egg file that is not necessarily related to any one node. (e.g. dynamic weather effect properties for a whole level, information to alter game settings like gravity, wind...)
2. Creating entity systems for mapping logic that some other game engines have. (like complex objects in-game that need to be configured, e.g. ammo lockers that have distinct refill cycle lengths and maximum ammo capacity)

**Comment by rdb** [comment]
Are you suggesting the ability to define custom tags like `<MyTag> ... { ... }` ?  Currently, the tag names are built into the grammar.

I am not sure what problem this solves that cannot be solved with the existing tag system.

**Comment by lifelandman** [comment]
Well, the limitations of tags that I'm complaining about is that they can only hold a single string value, and if you use tags to define custom properties of a model/map/level you have to search for them after the model has already loaded, as opposed to performing an operation when the egg-parser reaches that point of the file.

I was really thinking of something that would allow for a programmer to define their own scalars and handle all of those at once. something like this:
```
base.loader.loadModel('someEgg.egg', handlers = ['ammoBox', create_ammobox])#give loader a list of handlers for custom entries

def create_ammobox(name, **scalars):#the name of the entry and the scalars within get passed as named arguments
  if 'transform' in scalars:#\/ obviously panda3d wouldn't know what value a scalar is supposed to be so just give us the string value.
   strPos = scalars.get('transform').split(' ')
   pos = []
   i = 0
   while i >3:
    pos.append(float(strPos[i]))
  else: pos = [0,0,0]
  #imagine I did the same thing here for rotation and ammo capacity

newBox = ammoBox(pos, hpr, ammoCap)

base.thing_that_needs_know_of_amBoxes.addBox(newBox)

return newBox.geometry#give the loader a node to place in the loaded egg file, unless custom entry is abstract/not physical (i.e. gravity)
```

Of course, I understand that this is not a trivial feature and might be a pointless idea. the idea just kind of jumped into my head and I thought I should throw it out there.

**Comment by BMaxV** [comment]
I agree with rdb, the idea for such a system is probably good but I don't think this functionality should live in the model loading step.

**Comment by ArsThaumaturgis** [comment]
I suppose that my question is this: How would such tags be defined when creating the egg-file?

Of course, if the file is created in code, then the tags can be specified in code. But if the file is exported from a 3D editor, where would they be specified if _not_ on a node...?

**Comment by Wizzerinus** [comment]
That sounds like a question you should ask the exporting tool about. I bet if this functionality is possible and used, one can modify the exporting tool to write some kind of properties on the nodes (whether it's a i.e. blender plugin or an out-of-game postprocessing tool like panda-utils)

That said this functionality seems very hard to add considering the egg syntax reader is based on bison and it's a common thing for bison parsers to include the actual functionality at the same time. Not to mention this would require modifying the bam structure as well, and so on, for a non-obvious benefit - I'm not actually sure if the create_ammobox function provided is better than something like
```py
model = base.loader.loadModel('someEgg.egg')
for submodel in model.find("**/=postprocess-this"):
  tagName = submodel.getTag("postprocess-this")  # not sure on the syntax of this
  if tagName == "ammobox":  # in a real application this can just be a dictionary of different tagNames -> function that parses those tag names
    submodel.assign(create_ammobox(submodel))  # not sure on the syntax of this
``` considering the transform also is manually parsed there instead of being automatically applied by the scene graph.

**Comment by serkkz** [comment]
I'm wondering what prevents you from using serialization capabilities such as json instead of expanding egg? It's easier and it's not related to any limitations.

**Comment by lifelandman** [comment]
@ArsThaumaturgis The suggestion is for custom entries in egg files specifically, like
```
<AmmoBox> lighthouse_ammo_1 {
 <pos> { 19 23 45 }
 <hpr> { 52 0 0 }
}
```
There are also egg file entries like `<CoordinateSystem>` and `<Texture>` that are agnostic to geometry.

@Wizzerinus I imagine you're right that this functionality would be hard to add, especially since you seem to know more about the topic than me. I do want to note that I mostly suggested this as a way to differentiate egg files from larger, more long-term bam files and take advantage of their being text-format to improve the legibility of both egg files themselves and code when dealing with custom objects.

 After all, if you want your custom objects/entities to have parameters, you would have to either create new tags for these parameters, or have the tag key represent the custom object and parse through and translate a long string to the correct types to retrieve the parameters. This would be especially tough if you want to do something like the source engine's mapping logic, where objects/entities will communicate to each other through set inputs/outputs.

@serkkz I'm not yet familiar with json. I'll look into that now.
